src/ecies.cpp

```diff
@@ -0,0 +1,497 @@
+    size_t _symmetricKeySize = 0;
+    std::function<std::unique_ptr<SymmetricCipherI> (const std::vector<uint8_t> &)> _symmetricCipherFactoryFunction;
+    std::vector<uint8_t> _kdfSalt;
+    std::vector<uint8_t> _macSalt;
+    AsymmetricPublicKey _pubKey;
+    AsymmetricPrivateKey _privKey;
+
+};
+
+class ECIESImpl
+{
+public:
+    ECIESImpl(const ECIESCtxBuilder &ctxBuilder)
+        : kdf(ctxBuilder._impl->_kdf),
+          _macKeySize(ctxBuilder._impl->_macKeySize),
+          _macFactoryFunc(ctxBuilder._impl->_macFactoryFunction),
+          _mac(nullptr),
+          _symmetricKeySize(ctxBuilder._impl->_symmetricKeySize),
+          _symCipherFactoryFunc(ctxBuilder._impl->_symmetricCipherFactoryFunction),
+          _kdfSalt(ctxBuilder._impl->_kdfSalt),
+          _macSalt(ctxBuilder._impl->_macSalt),
+          _symcryptor(nullptr)    {}
+
+    virtual ~ECIESImpl()
+    {
+
+    }
+
+protected:
+    void createKeysAndInstantiateMacAndSymCipher(std::vector<uint8_t> sharedSecret)
+    {
+        utility::Finally sharedSecret_deleter([&sharedSecret]() {
+            utility::vectorCleanse(sharedSecret);
+        });
+
+        std::vector<uint8_t> result;
+        utility::Finally result_deleter([&result]() {
+            utility::vectorCleanse(result);
+        });
+
+        result = kdf->deriveKey(sharedSecret, _macKeySize + _symmetricKeySize, _kdfSalt);
+
+        std::vector<uint8_t> _key_encrypt(_symmetricKeySize);
+        utility::Finally key_encrypt_deleter([&_key_encrypt]() {
+            utility::vectorCleanse(_key_encrypt);
+        });
+
+        std::vector<uint8_t> _key_mac(_macKeySize);
+        utility::Finally key_mac_deleter([&_key_mac]() {
+            utility::vectorCleanse(_key_mac);
+        });
+
+        /* Copy first part (encryption key) */
+        std::copy(result.begin(), result.begin() + _symmetricKeySize, _key_encrypt.begin());
+
+        /* Copy second part (mac key) */
+        std::copy(result.begin() + _symmetricKeySize, result.end(), _key_mac.begin());
+
+        /* Write zeros into result */
+        utility::vectorCleanse(result);
+
+        /* We have calculated the MAC key now -> set it */
+        _mac = _macFactoryFunc(_key_mac);
+
+        /* initialize the de/encryptor */
+        _symcryptor = _symCipherFactoryFunc(_key_encrypt);
+
+    }
+
+    const std::shared_ptr<KeyDerivationFunction> kdf;
+    const size_t _macKeySize;
+    std::function<std::unique_ptr<MessageAuthenticationCode> (const std::vector<uint8_t> &)> _macFactoryFunc;
+    std::unique_ptr<MessageAuthenticationCode> _mac;
+    const size_t _symmetricKeySize;
+    std::function<std::unique_ptr<SymmetricCipherI>(const std::vector<uint8_t> &)> _symCipherFactoryFunc;
+    const std::vector<uint8_t> _kdfSalt;
+    const std::vector<uint8_t> _macSalt;
+
+    std::unique_ptr<SymmetricCipherI> _symcryptor;
+    std::vector<uint8_t> _mac_value;
+    bool _isFinished = false;
+};
+
+class ECIESEncryptionCtx::Impl : ECIESImpl
```

Comment:
Shouldn't this also be public inheritance like you did it for `ECIESDecryptionCtx`?

```suggestion
class ECIESEncryptionCtx::Impl : public ECIESImpl
```

Reply:
Done